[improve](schema-change) support nested type with varchar type to support length growing

amorynan opened the pull request:
### What problem does this PR solve?
cherry-pick from #46639
https://github.com/apache/doris/pull/48607
https://github.com/apache/doris/pull/49452 should be picked soon to resovle problem introduced by the pr 
Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

### Release note

None

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

Commit:
fix for branch-3.0

hello-stephen commented:
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

amorynan commented:
run buildall

Commit:
fix core

amorynan commented:
run buildall

Commit:
update cases

amorynan commented:
run buildall

Commit:
add cases

amorynan commented:
run buildall

amorynan committed:
Merge branch 'branch-3.0' into amory-pick

Hastyshell committed:
[fix](schema-change) Complete check for string type length change (#48607)

Related PR: introduce #46639

Problem Summary:

After the change of #46639, shorten length for CHAR type escapes from
checking of schema change. Fix in this PR and add some regression test
cases to verify it.

amorynan commented:
run buildall

Commit:
Merge branch 'branch-3.0' into amory-pick

amorynan commented:
run buildall

TangSiyang2001 commented on fe/fe-core/src/main/java/org/apache/doris/catalog/ColumnType.java:
```diff
@@ -168,6 +169,72 @@ static boolean isSchemaChangeAllowed(Type lhs, Type rhs) {
+    // return true if the checkType and other are both char-type otherwise return false,
+    // which used in checkSupportSchemaChangeForComplexType
+    private static boolean checkSupportSchemaChangeForCharType(Type checkType, Type other) throws DdlException {
+        if (checkType.isStringType() && other.isStringType()) {
+            checkForTypeLengthChange(checkType, other);
+            return true;
```
pick #49452 to fix problem here.

Currently nested types only support light schema change for internal fields, for string types, only varchar can do light schema change.

Commit:
fix cases

amorynan commented:
run buildall

Commit:
fix case

amorynan commented:
run buildall

Commit:
fix case out file

amorynan commented:
run buildall

dataroaring reviewed (approved):
LGTM
